### DISCUSSION TITLE
No clear way in how to create an emqtt OPT client using emqtt with quick over mqtt

### DISCUSSION THREAD

**Discussion by SaveMe0** [emqx/emqx#14452] [Q&A]
I've created a new discussion because the previous one https://github.com/emqx/emqx/discussions/14415#top was getting to cluttered to address my issues using emqtt. I sincerely hope I don't annoy the maintainer and the contributors to he emqtt application. 
In this discussion I want to point the results of using these functions in the emqtt client I currently building
   quic_connect/1
   open_quic_connection/1
   quic_mqtt_connect/1

The following snippet is part of the module I've create you will find attached to this discussion. In this snippet the connPid is return value of emqtt:start_link/1.

```
cnctState(cast, cnct, Data) ->
    io:format("My proccess Pid ~p~n", [self()]),
	io:format("ConnPid ~p~n", [Data#comm.connPid]),
    ConnPid = Data#comm.connPid, 
    % emqtt:quic_connect(ConnPid),
    % emqtt:open_quic_connection(ConnPid),
    % emqtt:quic_mqtt_connect(ConnPid),
    {next_state, subState, Data};
cnctState(EventType, EventContent, Data) ->
    handle_event(EventType, EventContent, Data).   
```

Result of using emqtt:quic_connect(ConnPid)

```
1> com2:start().
ConnPid <0.216.0>
My proccess Pid <0.215.0>
ok
2> com2:cnct().
My proccess Pid <0.215.0>
ok
ConnPid <0.216.0>
3> is_process_alive(<0.216.0>).
false
4> 
```

Result of using emqtt:open_quic_connection(ConnPid)
```
1> com2:start().
ConnPid <0.216.0>
My proccess Pid <0.215.0>
ok
2> com2:cnct().
My proccess Pid <0.215.0>
ok
ConnPid <0.216.0>
3> is_process_alive(<0.216.0>).
true
4
```
Result of using emqtt:quic_mqtt_connect(ConnPid)
```
1> com2:start().
ConnPid <0.216.0>
My proccess Pid <0.215.0>
ok
2> com2:cnct().
My proccess Pid <0.215.0>
ok
ConnPid <0.216.0>
=ERROR REPORT==== 19-Dec-2024::17:54:50.945500 ===
    state: initialized
    msg: unexpected_event
    clientid: <<>>
    event: quic_mqtt_connect
    event_type: {call,{<0.215.0>,#Ref<0.79571594.3575906305.111200>}}
3> is_process_alive(<0.216.0>).
true
```
As you can see from the three tests above 
```
calling function emqtt:quic_connect(ConnPid)   
        kills the emqtt process 

calling function emqtt:open_quic_connection(ConnPid) 
       does not kill the emqtt process

calling function emqtt:quic_mqtt_connect(ConnPid) 
       does not kill the emqtt process the process
       generates a msg: unexpected_event
       an empty clientid: <<>> is empty. 
```

For the last function call  emqtt:quic_mqtt_connect(ConnPid) If I had created a clientID it would have been shown. If you would like to create your own this link explains how [](https://www.hivemq.com/blog/supporting-mqtt-client-identifiers-23-characters/)
I've been informed that the  generation of a ConnPid is dependent on a ClientID and for this reason it automatically generated by emqtt. Perhaps it is stored somewhere in emqx,  but it certainly does not show up in emqtt's Option list where I'm assuming it should be.

In my next tests I will show the results of subscribing topics from the broker. I can start with subscriptions because I have a EK9160 Bus Coupler that automatically publishes data to the broker from devices connected to its input slaves. I have conducted tests with connection functions emqtt:open_quic_connection(ConnPid)  and emqtt:quic_mqtt_connect(ConnPid)  since in my case they don't kill the emqtt process when connecting to the broker. Here is the snippet of code I will be using for his test

```
subState({call, From}, sub, Data) ->
    io:format("CommPid ~p~n", [Data#comm.connPid]),
    ConnPid = Data#comm.connPid,
    %Properties = #{},
    Properties = #{ 'Maximum-Packet-Size' => 1048576,
                    'Receive-Maximum' => 32,
                    'Retain-Available' => 1,
                    'Shared-Subscription-Available' => 1,
                    'Subscription-Identifier-Available' => 1,
                    'Topic-Alias-Maximum' => 65535,
                    'Wildcard-Subscription-Available' => 1,
                    'Payload-Format-Indicator' => 1,
                    'Message-Expiry-Interval' => 10,
                    'Content-Type' => <<"Json">>,
                    'Server-Keep-Alive' => 30,
                    'Maximum-QoS' => 1 },

    SubOpts = [{qos, 1}],
    Subscriptions = [{<<"EK9160/EK-98EFB6/Stream1/Json/Tx/Data">>, SubOpts}],
    Reply = emqtt:subscribe(ConnPid, Properties, Subscriptions),
    io:format("Subscription~p~n", [Reply]),
    {keep_state,Data, [{reply, From, Data}]};
subState(EventType, EventContent, Data) ->
    handle_event(EventType, EventContent, Data). 

```
 Subscription test after connecting with emqtt:open_quic_connection(ConnPid) 
```
1> com2:start().
ConnPid <0.216.0>
My proccess Pid <0.215.0>
ok
2> com2:cnct().
My proccess Pid <0.215.0>
ok
ConnPid <0.216.0>
3> com2:sub().
CommPid <0.216.0>
=ERROR REPORT==== 19-Dec-2024::19:04:55.689142 ===
    state: initialized
    msg: unexpected_event
    clientid: <<>>
    event: {subscribe,#{'Content-Type' => <<"Json">>,
                        'Server-Keep-Alive' => 30,'Retain-Available' => 1,
                        'Maximum-Packet-Size' => 1048576,
                        'Receive-Maximum' => 32,
                        'Shared-Subscription-Available' => 1,
                        'Subscription-Identifier-Available' => 1,
                        'Topic-Alias-Maximum' => 65535,
                        'Wildcard-Subscription-Available' => 1,
                        'Payload-Format-Indicator' => 1,
                        'Message-Expiry-Interval' => 10,'Maximum-QoS' => 1},
                      [{<<"EK9160/EK-98EFB6/Stream1/Json/Tx/Data">>,
                        #{nl => 0,qos => 1,rap => 0,rh => 0}}]}
    event_type: {call,{<0.215.0>,#Ref<0.390911650.1707868161.239416>}}
```

Subscription test after connecting with emqtt:quic_mqtt_connect(ConnPid)
```
1> com2:start().
ConnPid <0.216.0>
My proccess Pid <0.215.0>
ok
2> com2:cnct().
My proccess Pid <0.215.0>
ok
ConnPid <0.216.0>
=ERROR REPORT==== 19-Dec-2024::19:10:37.455337 ===
    state: initialized
    msg: unexpected_event
    clientid: <<>>
    event: quic_mqtt_connect
    event_type: {call,{<0.215.0>,#Ref<0.1017453540.3587702785.207124>}}
3> com2:sub().
```
Both test fail with an error report and then hang. Of the two test however the one using emqtt:open_quic_connection(ConnPid) seems to be working better than the one with emqtt:quic_mqtt_connect(ConnPid). In the one with emqtt:open_quic_connection(ConnPid) it could be that the data is being published but I don't know how to access it to print it. 

I've conducted these tests with the hope that someone in the eqmtt team can suggest ways in how I correct my subscription code.

Thank you
[com2.zip](https://github.com/user-attachments/files/18201883/com2.zip)

**Comment by qzhuyan** [comment]
here is the minimal stuff you want, assuming you are connecting to "127.0.0.1:14567"

```
{ok, Pid} = emqtt:start_link([{host, {127,0,0,1}},
                           {port, 14567}
                          ]),
 _ =emqtt:quic_connect(Pid),
emqtt:subscribe(Pid, <<"a/b/c">>, qos0).
```

**Reply by SaveMe0** [reply]
Hi William, 
yes of course I know that - honestly. 

Perhaps the reason why I the emqtt:quic_connect/1 function is not working for me, is because I might be missing some dependencies in rebar.config file.  Here is a copy of my rebar.conf file
```
{erl_opts, [debug_info]}.
{deps, [{emqtt, {git, "https://github.com/emqx/emqtt.git", {branch, "master"}}}]}.

{shell, [
    %% {config, "config/sys.config"},
    {apps, [newapp]}
]}.
```
Could you be so kind to check it over, and if I'm missing something, post them in this discussion. Thank you.

**Reply by SaveMe0** [reply]
Also William,
has emqtt with quic over mqtt been tested for Erlang/OTP 27?
I'm running  emqtt with  Erlang/OTP 27 [erts-15.1.2] [source] [64-bit] [smp:1:1] [ds:1:1:10] [async-threads:1] [jit:ns] on Debian 12.
